pandas/src/algos_rank_helper.pxi.in

```diff
@@ -0,0 +1,392 @@
+                        ranks[argsorted[j]] = 2 * i - j - dups + 2
+                elif tiebreak == TIEBREAK_DENSE:
+                    total_tie_count += 1
+                    for j in range(i - dups + 1, i + 1):
+                        ranks[argsorted[j]] = total_tie_count
+                sum_ranks = dups = 0
+    {{endif}}
+    if pct:
+        return ranks / count
+    else:
+        return ranks
+
+
+def rank_2d_{{name}}(object in_arr, axis=0, ties_method='average',
+                     ascending=True, na_option='keep', pct=False):
+    """
+    Fast NaN-friendly version of scipy.stats.rankdata
+    """
+
+    cdef:
+        Py_ssize_t i, j, z, k, n, dups = 0, total_tie_count = 0
+
+        {{if name == 'generic'}}
+        Py_ssize_t infs
+        {{endif}}
+
+        ndarray[float64_t, ndim=2] ranks
+        {{if name == 'int64' or name == 'uint64'}}
+        ndarray[{{ctype}}, ndim=2, cast=True] values
+        {{else}}
+        ndarray[{{ctype}}, ndim=2] values
+        {{endif}}
+
+        ndarray[int64_t, ndim=2] argsorted
+
+        {{if name == 'uint64'}}
+        {{ctype}} val
+        {{else}}
+        {{ctype}} val, nan_value
+        {{endif}}
+
+        float64_t sum_ranks = 0
+        int tiebreak = 0
+        bint keep_na = 0
+        float count = 0.0
+
+    tiebreak = tiebreakers[ties_method]
+
+    keep_na = na_option == 'keep'
+
+    in_arr = np.asarray(in_arr)
+
+    if axis == 0:
+        values = in_arr.T.copy()
+    else:
+        values = in_arr.copy()
+
+    {{if name == 'generic'}}
+    if values.dtype != np.object_:
+        values = values.astype('O')
+    {{endif}}
+
+    {{if name != 'uint64'}}
+    if ascending ^ (na_option == 'top'):
+        nan_value = {{pos_nan_value}}
+    else:
+        nan_value = {{neg_nan_value}}
+
+    {{if name == 'generic'}}
+    mask = lib.isnullobj2d(values)
+    {{elif name == 'float64'}}
+    mask = np.isnan(values)
+    {{elif name == 'int64'}}
+    mask = values == iNaT
+    {{endif}}
+
+    np.putmask(values, mask, nan_value)
+    {{endif}}
+
+    n, k = (<object> values).shape
+    ranks = np.empty((n, k), dtype='f8')
+
+    {{if name == 'generic'}}
+    try:
+        _as = values.argsort(1)
+    except TypeError:
+        values = in_arr
+        for i in range(len(values)):
+            ranks[i] = rank_1d_generic(in_arr[i],
+                                       ties_method=ties_method,
+                                       ascending=ascending,
+                                       pct=pct)
+        if axis == 0:
+            return ranks.T
+        else:
+            return ranks
+    {{else}}
+    if tiebreak == TIEBREAK_FIRST:
+        # need to use a stable sort here
+        _as = values.argsort(axis=1, kind='mergesort')
+        if not ascending:
+            tiebreak = TIEBREAK_FIRST_DESCENDING
+    else:
+        _as = values.argsort(1)
+    {{endif}}
+
+    if not ascending:
+        _as = _as[:, ::-1]
+
+    {{if name == 'generic'}}
```

Comment:
maybe change generic -> object so this is not necessary? (in a higher level definition)